init-service/src/main/java/uk/ac/ox/ndph/mts/init_service/service/SiteServiceInvoker.java

```diff
@@ -3,59 +3,51 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Value;
-import org.springframework.http.HttpHeaders;
-import org.springframework.http.MediaType;
 import org.springframework.stereotype.Service;
 import org.springframework.web.reactive.function.client.WebClient;
-import reactor.core.publisher.Mono;
 import uk.ac.ox.ndph.mts.init_service.exception.DependentServiceException;
 import uk.ac.ox.ndph.mts.init_service.exception.NullEntityException;
 import uk.ac.ox.ndph.mts.init_service.model.Entity;
+import uk.ac.ox.ndph.mts.init_service.model.IDResponse;
 import uk.ac.ox.ndph.mts.init_service.model.Site;
 
+import java.util.ArrayList;
 import java.util.List;
 
 @Service
-public class SiteServiceInvoker implements ServiceInvoker {
+public class SiteServiceInvoker extends ServiceInvoker {
     private static final Logger LOGGER = LoggerFactory.getLogger(SiteServiceInvoker.class);
 
     @Value("${site.service}")
     private String siteService;
 
-    private final WebClient webClient;
+    public SiteServiceInvoker() { }
+    public SiteServiceInvoker(WebClient webClient) { super(webClient); }
 
-    public SiteServiceInvoker() {
-        this.webClient = WebClient.create(siteService);
-    }
-    public SiteServiceInvoker(WebClient webClient) {
-        this.webClient = webClient;
-    }
-
-    public Site send(Entity site) throws DependentServiceException {
+    protected String create(Entity site) throws DependentServiceException {
 
         try {
-            return webClient.post()
-                    .uri(siteService + "/sites")
-                    .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
-                    .body(Mono.just(site), Site.class)
-                    .retrieve()
-                    .bodyToMono(Site.class)
-                    .block();
+            String uri = siteService + "/sites";
+            IDResponse responseFromSiteService = sendBlockingPostRequest(uri, site, IDResponse.class);
+            return responseFromSiteService.getId();
+
         } catch (Exception e) {
             LOGGER.info("FAILURE siteService {}", e.getMessage());
             throw new DependentServiceException("Error connecting to site service");
         }
     }
 
-    public void execute(List<Site> sites) throws NullEntityException {
+    public List<String> execute(List<Site> sites) throws NullEntityException {
```

Comment:
Could we create a generic execute method using <Entity> as this code is duplicated in other invokers?

Reply:
Thanks Darren.   I've had a go at this but not sure whether it improves code clarity or not, on balance.   The execute methods were all a little different (the one in SiteServiceInvoker returns the site IDs to the caller, the one in RoleServiceInvoker returns void since those IDs aren't needed, and the one in PractitionerServiceInvoker does a bit more work with a second rest call).   I've had a go at extracting a method which will work for the site and role cases, leaving the practitioner case on its own.  That leaves the method both overloaded and overridden, which is a bit trixy for the future reader.  wdyt?

Reply:
yeah thats the approach I took when I played with it before adding the comment - you could change List<String> siteIDs = new ArrayList<>(); to List<String> entityIDs = new ArrayList<>(); or just List<String> IDs = new ArrayList<>();

Reply:
Good point that's the wrong name - fixed.